primitives/src/crh/poseidon/mod.rs

```diff
@@ -1,427 +1,213 @@
-        // in that case compute the inverses individually
-        if accum_prod == P::Fr::zero() {
-            for i in 0..(vec_state.len() - 1) {
-                if vec_state[i][0] != P::Fr::zero() {
-                    vec_state[i][0] = vec_state[i][0].inverse().unwrap();
-                }
-            }
-        } else {
-
-            // Calculate the inversion of the products
-            // Use Montgomery simultaneous inversion
-            let mut w_bar = accum_prod.inverse().unwrap();
-
-            // Extract the individual inversions
-            let mut idx: i64 = w.len() as i64 - 2;
-            for i in (0..vec_state.len()).rev() {
-                let vec_1 = vec_state[i][0].clone();
-                vec_state[i][0] = w_bar * &w[idx as usize];
-                w_bar = w_bar * &vec_1;
-                idx -= 1;
-            }
+    #[inline]
+    fn get_hash(mut state: Vec<F>, inputs: Vec<F>) -> F {
+        for (input, s) in inputs.iter().zip(state.iter_mut()) {
+            *s += input;
         }
+        Self::poseidon_perm(&mut state);
+        state[0]
     }
 
-    fn poseidon_perm_gen(vec_state: &mut Vec<Vec<P::Fr>>) {
-
-        // index that goes over the round constants
-        let mut round_cst_idx: usize = 0;
-
-        // Full rounds
-        for _i in 0..P::R_F {
-            Self::poseidon_full_round(vec_state, &mut round_cst_idx);
-
-            // Perform the matrix mix
-            for i in 0..vec_state.len() {
-                //Self::matrix_mix_short(&mut vec_state[i]);
-                Self::matrix_mix_short(&mut vec_state[i]);
+    #[inline]
+    // Padding strategy is described in https://eprint.iacr.org/2019/458.pdf(Section 4.2)
+    fn pad_and_finalize(&self) -> F
+    {
+        // Constant input length instance
+        if self.input_size.is_some() {
+
+            // The constant size is modulus rate, so we already have the hash output in state[0]
+            // as a permutation is applied each time pending reaches P::R length
+            if self.pending.is_empty() {
+                self.state[0].clone()
             }
 
+            // Pending is not empty: pad with 0s up to rate then compute the hash,
+            else {
+                Self::get_hash(self.state.clone(), self.pending.clone())
+            }
         }
 
-        // Partial rounds
-        for _i in 0..P::R_P {
-            Self::poseidon_partial_round(vec_state, &mut round_cst_idx);
+        // Variable input length instance
+        else {
 
-            // Perform the matrix mix
-            for i in 0..vec_state.len() {
-                //Self::matrix_mix_short(&mut vec_state[i]);
-                Self::matrix_mix_short(&mut vec_state[i]);
+            // The input is of variable length, but always modulus rate: result is already available
+            // in state[0] as a permutation is applied each time pending reaches P::R length
+            if self.mod_rate {
+                self.state[0].clone()
```

Comment:
As discussed before, this way of selecting the padding mode does allow trivial collisions.
If one wants to be efficient, then one can set up a fixed-length hash instead, or separately instantiate a sponge that only allows inputs that are a multiple of the rate.